rpm.spec: Fix mock build, requires and quoting

sshambar opened the pull request:
This patch fixes rpmbuild/mock failures due to missing BuildRequires,
and references to absolute filenames during install.  In addition
it corrects the quoting of the bridge-stp file contents, and
correct several macro references.

Now mock builds correctly on Fedora/CoreOS etc

Signed-off-by: Scott Shambarger <devel@shambarger.net>

commodo commented on rpm.spec:
```diff
@@ -27,18 +28,17 @@ on Linux bridges when MSTP is used.
-mkdir -p %{buildroot}/%{_sysconfdir}/sysconfig/network-scripts
-cat <<END > %{buildroot}/%{_sysconfdir}/sysconfig/network-scripts/bridge-stp
+mkdir -p %{buildroot}%{_sysconfdir}/sysconfig/network-scripts
+cat <<"END" > %{buildroot}%{_sysconfdir}/sysconfig/network-scripts/bridge-stp
```
are you sure that `END` needs quoting?
i'm a little paranoid about that;

sshambar replied:
Yes, "END" means the content of the here-doc isn't parsed for $VAR refs and "`" executables... the old version tried to run "man mstpctl-utils-interfaces" and placed the output (empty) in the resulting file.  "END" means all the content is included as is.

commodo commented:
rest looks good

i'd let @PaulSD have a look as well;

if no objections, i will merge this sometime next week;

PaulSD commented:
Two comments:
1) I would prefer `cat <<'END'` (single quotes instead of double quotes), as it makes it slightly clearer that the block shouldn't be expanded (since double quotes usually imply expansion).
2) I'm not sure what the purpose of the systemd unit file is ... `bridge-stp` automatically starts/stops `mstpd` automatically as needed (even if you are not using the documented `mstp_config_bridge` script and are using some other mechanism to configure the bridge), and systemd process monitoring doesn't help in this case (eg. it can't auto-recover from an mstpd crash since mstpd must be reconfigured after it is restarted and the systemd unit doesn't do that).  If there is some actual use case for it that I am not seeing, then I am OK with adding it to the RPM as is done here so that it is available to users who want it, but maybe we should add a comment to the bridge-stp.conf file to note that the systemd unit isn't used?  Otherwise, I would just leave it out to avoid confusing users.

sshambar commented:
>     1. I would prefer `cat <<'END'` 
Sounds good, no difference between the two :)

>     2. I'm not sure what the purpose of the systemd unit file is ... 
I think the systemd unit has some use as it allows easy control of the daemon that's not easy with auto-spawning (such as limiting capabilities, containerizing, general hardening) for those who'd like the option via drop-ins.  In addition, it makes for very easy place to check for status/logging.

I agree that it should probably not have the "Restart" as that wouldn't work -- however it is a nice place to add a hook to reconfigure the bridge if the daemon does fail, or is killed!

(As an aside) it actually might be nice to have a "graceful restart" option to the daemon so an upgrade could replace it - possibly serializing the config in a tmpfile and reading that on restart, but that's a project for another day :D.

Even "if" the unit file remains installed, I think it should be disabled by default, and just there in case someone wants to use it (possibly with a note indicating that restarting it would fail!)

Let me know your thoughts...

PaulSD commented:
Good points about capabilities/logging/etc ... I'm OK with installing the systemd unit and leaving it disabled by default.

Feel free to change the quoting and add some comments (You can `git commit -a --amend ; git push --force` to update the PR) ... Or I can do it after this is merged.

sshambar commented:
OK, made the quoting change, but also a few other fixes:

 - git mv'd rpm.spec to mstpd.spec so that rpmbuild -ta <tar> works
 - added Source: to ref github (the official source, as recommended for the Source entry)
 - updated make rpm to stop on errors, only include the source (not .git), and use .tar.gz so that the Source: link works if used.
 - updated the mstpd.service file so that now restart/reload work correctly (still disabled by default), and made it before network-pre.target (before bridge-stp may be called)

In the process of testing the changes, I found a few bugs unrelated to the spec/service fixes, and was going to include fixes, but I think since they're orthogonal issues, you'd prefer they have they're own Issues/PullReqs...

commodo commented:
will wait for @PaulSD to take a look as well;

commodo commented:
continuing here;
i can add a Fedora build in the [build.yml file](https://github.com/mstpd/mstpd/blob/master/.github/workflows/build.yml)

we can add as many docker images as needed;
i never thought about build testing this on something other than Ubuntu;
but Fedora may make sense

sshambar commented:
Few updates that are specific to RPM based systems:

 - added a NetworkManager dispatcher script, to the bridge if correctly configured after nmcli con up on those systems
 - fixed replacement ifquery to fail on non-configured interfaces
 - updated sample interfaces file in .spec to comment out port-based interfaces so they don't create warnings when invalid, and updates their values to make it clear they refer to port names

I can make the dispatcher addition and ifquery fix independent pulls, but since they only applies to platforms that use rpm, it seemed like they belong together :)

sshambar commented:
> i can add a Fedora build in the [build.yml file](https://github.com/mstpd/mstpd/blob/master/.github/workflows/build.yml)

I'll think I'll leave that to you... I'm not really sure which Fedora image to use, and I don't know if you'd like to build the rpm as well :)

PaulSD commented:
Sorry it is taking a while for me to review again - I can't seem to catch a break this week.  I'll try to look this weekend.

PaulSD commented on mstpd.spec:
```diff
@@ -27,18 +32,21 @@ on Linux bridges when MSTP is used.
-sed -i -e 's|/etc/network/interfaces|/etc/sysconfig/network-scripts/bridge-stp|g' %{buildroot}/%{_libexecdir}/mstpctl-utils/mstp_config_bridge
+mkdir -p $RPM_BUILD_ROOT%{_prefix}/lib/NetworkManager/dispatcher.d
+install -m 755 -p utils/nm-dispatcher \
+  $RPM_BUILD_ROOT%{_prefix}/lib/NetworkManager/dispatcher.d/45-mstpd
```
Shouldn't this be `%{buildroot}` instead of `$RPM_BUILD_ROOT`?

sshambar replied:
They're the same, but RedHat packages seem to prefer $RPM_BUILD_ROOT in their install scripts...

PaulSD replied:
Hmmm... Yeah, now that you point it out, I see that they do seem to prefer that.  In that case, let's change the rest of the `%{buildroot}` macros we are currently using in this section to `$RPM_BUILD_ROOT`.

PaulSD commented on utils/nm-dispatcher.in:
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/sh
```
Should be `/bin/sh`, not `/usr/bin/sh`

sshambar replied:
/usr/bin/sh is actually the preferred shell, and the rpm build will actually change it if it's /bin/sh...

PaulSD replied:
Bleh ... It looks like this is new behavior related to the python2->python3 migration in Fedora/RHEL?  I can't find anything specifically documenting the sh path change or the reasoning behind it, although I see they made several changes to this between RHEL6, RHEL7, and RHEL8.

There are two problems here:
1) This script would be useful on non-RPM systems (at least Debian-based systems), so I would like this script to be usable on both (even if we aren't currently packaging it for Debian - we can add that bit later).  But Debian-based systems don't have /usr/bin/sh, so this script won't work on them as-is.  If `rpmbuild` rewrites the shebang automatically, I'm thinking maybe we should leave it /bin/sh in the repo for use on non-RPM systems, and let `rpmbuild` rewrite it for RPMs.
2) I'm really trying to keep things consistent, and we currently have several scripts in mstpd that use /bin/sh ... So if we're going to change it, I think we should change all of the scripts ... But (circling back to the previous problem) changing all of the scripts breaks other systems.  So, again, I'm thinking let's make this /bin/sh and let `rpmbuild` change it if needed.

PaulSD commented on utils/nm-dispatcher.in:
```diff
@@ -0,0 +1,19 @@
+
+config_cmd='@configbridgefile@'
+
+if [ "$action" = "up" -a -n "$interface" ]; then
```
Should use `[ ... ] && [ ... ]` instead of `[ ... -a ... ]`.  I don't remember the exact reasoning for this - POSIX compliance or something - but the rest of the mstpd scripts do it that way, so let's do that for consistency if nothing else.

sshambar replied:
Fixed... consistency is good :)

PaulSD reviewed (commented):
Some minor nits, but overall this is great work!

We should also adjust the documentation to cover use with networkmanager and/or the systemd script, but I think it is fine to merge this without documentation for the new bits.
